src/Lean/Compiler/IR/EmitLLVM.lean

```diff
@@ -0,0 +1,1521 @@
+  let xv ← emitLhsVal builder x
+  if n != 1
+  then throw (Error.compileError "expected n = 1 for emitDec")
+  else callLeanRefcountFn builder (kind := RefcountKind.dec) (checkRef? := checkRef?) xv
+
+def emitNumLit (builder : LLVM.Builder llvmctx)
+  (t : IRType) (v : Nat) : M llvmctx (LLVM.Value llvmctx) := do
+  if t.isObj then
+    if v < UInt32.size then
+      callLeanUnsignedToNatFn builder v
+    else
+      callLeanCStrToNatFn builder v
+  else
+    LLVM.constInt (← toLLVMType t) (UInt64.ofNat v)
+
+def toHexDigit (c : Nat) : String :=
+  String.singleton c.digitChar
+
+-- TODO(bollu) : Setup code sharing between 'EmitC' and 'EmitLLVM'
+def quoteString (s : String) : String :=
+  let q := "\"";
+  let q := s.foldl
+    (fun q c => q ++
+      if c == '\n' then "\\n"
+      else if c == '\r' then "\\r"
+      else if c == '\t' then "\\t"
+      else if c == '\\' then "\\\\"
+      else if c == '\"' then "\\\""
+      else if c.toNat <= 31 then
+        "\\x" ++ toHexDigit (c.toNat / 16) ++ toHexDigit (c.toNat % 16)
+      -- TODO(Leo) : we should use `\unnnn` for escaping unicode characters.
+      else String.singleton c)
+    q;
+  q ++ "\""
+
+def emitSimpleExternalCall (builder : LLVM.Builder llvmctx)
+  (f : String)
+  (ps : Array Param)
+  (ys : Array Arg)
+  (retty : IRType)
+  (name : String) : M llvmctx (LLVM.Value llvmctx) := do
+  let mut args := #[]
+  let mut argTys := #[]
+  for (p, y) in ps.zip ys do
+    if !p.ty.isIrrelevant then
+      let (_yty, yv) ← emitArgVal builder y ""
+      argTys := argTys.push (← toLLVMType p.ty)
+      args := args.push yv
+  let fnty ← LLVM.functionType (← toLLVMType retty) argTys
+  let fn ← LLVM.getOrAddFunction (← getLLVMModule) f fnty
+  LLVM.buildCall2 builder fnty fn args name
+
+-- TODO: if the external call is one that we cannot code generate, give up and
+-- generate fallback code.
+def emitExternCall (builder : LLVM.Builder llvmctx)
+  (f : FunId)
+  (ps : Array Param)
+  (extData : ExternAttrData)
+  (ys : Array Arg) (retty : IRType)
+  (name : String := "") : M llvmctx (LLVM.Value llvmctx) :=
+  match getExternEntryFor extData `c with
+  | some (ExternEntry.standard _ extFn) => emitSimpleExternalCall builder extFn ps ys retty name
+  | some (ExternEntry.inline "llvm" _pat)     => throw (Error.unimplemented "unimplemented codegen of inline LLVM")
+  | some (ExternEntry.inline _ pat)     => throw (Error.compileError s!"cannot codegen non-LLVM inline code '{pat}'")
+  | some (ExternEntry.foreign _ extFn)  => emitSimpleExternalCall builder extFn ps ys retty name
+  | _ => throw (Error.compileError s!"failed to emit extern application '{f}'")
+
+def getFunIdTy (f : FunId) : M llvmctx (LLVM.LLVMType llvmctx) := do
+  let decl ← getDecl f
+  let retty ← toLLVMType decl.resultType
+  let argtys ← decl.params.mapM (fun p => do toLLVMType p.ty)
+  LLVM.functionType retty argtys
+
+/-
+Create a function declaration and return a pointer to the function.
```

Comment:
Is there a reason for switching between multiline comment style? Further above you have multiple `--`. If you inteded for this to be a doc string that gets rendered by e.g. doc-gen4 you'll have to put `/--`

Reply:
Thanks! I did intend for it to be a docstring.